Initial Commit For api-03-fetch Objective

MarvelProgramming opened the pull request:
For issue #41

Commit:
Fixed Exclamation Visibility and Door Interaction Fix

-Made it so the door has it's "interactable" property set to "false" upon opening to prevent the exclamation
point from showing up and the door being set to the player's "inRangeObject"
-Made it so the door has it's "spellable" property set to "false" upon opening to prevent the player from
attempting to open the door and having an animation play as though they were unlocking it

Commit:
Forced Player Exclamation Point to Hide

-Set player's inRangeObject to null after opening the door, effectivly forcing the exclamation point to be hidden.
This makes it so the player doesn't need to walk out of and back in to the door's "interactable area" to see that
it can no longer be acted upon

Commit:
Merge pull request #15 from TwilioQuest/14-exclamation-point-remains-in-catacombs-doorway-even-after-opening-the-door

Fixed Exclamation Visibility and Door Interaction Fix

Commit:
Initial Commit for api-03-fetch Objective

Commit:
Updated Objective

-Updated description.md, example.js, objective.json, and validator.js based on new direction for objective

Commit:
Updated Objective

-Added first draft of description
-Updated validator to use response from Web API as expected output, instead of using a hardcoded value
-Updated example.js verbiage and Web API endpoint

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 41-api-03-fetch-fetch-from-twilioquest-api

Commit:
Moved Objective Into lovelace_tower

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 41-api-03-fetch-fetch-from-twilioquest-api

Commit:
Updated Objective

-Updated objective.json such that the player must paste the magical phrase they get back from the api, instead of returning it from their function
-Updated description, objective.json, and validator to account for the new direction of the objective (this includes verbiage, variable names, etc)
-Updated the validator's "DIVINATION_API_ENDPOINT" reference to pull from the config file, instead of defining it locally

Commit:
Added Walkthrough

Commit:
Minor Updates To Walkthrough

Commit:
Updated Objective

-Updated description, objective.json, and validator verbiage
-Replaced "DIVINATION_API_ENDPOINT" reference with "MAGIC_API_ENDPOINT"

joenash commented on levels/lovelace_tower/objectives/api-03-fetch/walkthrough.md:
```diff
@@ -0,0 +1,63 @@
+# What is fetch?
+
+Sometimes we want to get data or perform an action on another computer. One common way to do this is by visiting websites. Whenever you visit a website, your computer sends a request. The request's configuration helps determine what happens when the other device receives it. Javascript provides the `fetch` method, which allows you to send such requests to other computers based on a [URL](https:// ...
```
```suggestion
Sometimes we want to get data or perform an action on another computer. One common way to do this is by visiting websites. When you enter a the URL for a website in your web browser, the browser sends a HTTP request to the computer running the website, called the "server". The configuration of a HTTP request tells the server what to do with that request. Usually, when we visit a website from our browser, we send a HTTP GET request, which is a type of request which tells the server we want to "get" the website. 

To send HTTP requests, Javascript provides the `fetch` method. `fetch` lets us configure and send HTTP requests, including those that would let us get the contents of a website from our code! When we send a request with `fetch`, just like in the browser, we use a [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL).
```

If I'm not mistaken, they've learned about HTTP by this point, so I think we can be explicit about these requests

joenash commented on levels/lovelace_tower/objectives/api-03-fetch/walkthrough.md:
```diff
@@ -0,0 +1,63 @@
+HTTP Verbs (i.e. methods) represent the types of requests you can make. The most commonly used methods are `GET`, `POST`, `PATCH`, `PUT`, and `DELETE`.
+
+- `GET` asks for some sort of data or information from the other device
+- `POST` asks for something to be created by the other device
```
This is true for REST APIs, but probably not in general. Perhaps it should be:

```suggestion
- `POST` sends data to the other device
```

MarvelProgramming replied:
Very true and something I didn't fully consider! I'll comb through this and the other objectives and reword things a bit

joenash commented on levels/lovelace_tower/objectives/api-03-fetch/walkthrough.md:
```diff
@@ -0,0 +1,63 @@
+- `PUT` asks for something that exists already to be replaced entirely by the other device with what's being sent in the request
+- `PATCH` asks for something that exists already to be updated by the other device using a set of instructions being sent in the request
+- `DELETE` asks for something to be deleted by the other device
+
```
Might be worth mentioning that a response should still be sent back by all the requests, not just GET, and those responses can contain data just like GET

joenash commented on levels/lovelace_tower/objectives/api-03-fetch/walkthrough.md:
````diff
@@ -0,0 +1,63 @@
+console.log(response);
+```
+
+This example is similar to the last, only this time we're making a `POST` request with another option in our configuration object. This type of request is one that needs data to be sent along with it, and we accomplish that with the `body` property. The example is requesting that another computer create something new using the `Hello World` data we supplied.
````
Similar comment to above, that POST being for creating something new is probably fairly REST specific language

joenash reviewed (commented):
Looking awesome! Summary: we can probably talk in terms of HTTP from the off, since they've met it in part 1. Other parts are just some thoughts about how we talk about the HTTP methods and aren't a big deal/can be ignored

MarvelProgramming committed:
Update levels/lovelace_tower/objectives/api-03-fetch/walkthrough.md

Updated walkthrough verbiage

Co-authored-by: Joe  <1790822+joenash@users.noreply.github.com>

MarvelProgramming committed:
Update levels/lovelace_tower/objectives/api-03-fetch/walkthrough.md

Updated walkthrough verbiage

Co-authored-by: Joe  <1790822+joenash@users.noreply.github.com>

Commit:
Updated Walkthrough Based On Feedback

rmkubik commented on levels/lovelace_tower/objectives/api-03-fetch/example.js:
```diff
@@ -0,0 +1,9 @@
+(async function () {
+  console.log("Test case data:");
+  console.log((await getMagicalPhrase()).toString());
+})();
```
Probably good to avoid the IIFE here in favor of a `runTests` function declaration that we invoke on the next line .

nmadadair commented:
> _The House Gauntlet welcomes you to the third challenge of Lovelace Tower. In this challenge, you must test the powers of `await` by fetching a magical phrase. You may find this process helpful, Operator, when generating the Divination Spell that will be granted to you in Lovelace Tower upon completion of its challenges._

Would like some thoughts here on connecting fetch to creating the Divination Spell -- is the player still making this spell? In theory Obj 4 should grant them the spell since they are using it to open the door to the library. @MarvJohnson Have I missed something here?

EDIT: if we are no longer creating the Divination Spell in obj 4, we might say something about how fetch is part of the process of how spells are created (in general)?

Commit:
Merge branch 'milestone-1-aa3' of github.com:TwilioQuest/twilioquest-api-academy into 41-api-03-fetch-fetch-from-twilioquest-api

MarvelProgramming commented:
> Would like some thoughts here on connecting fetch to creating the Divination Spell -- is the player still making this spell? In theory Obj 4 should grant them the spell since they are using it to open the door to the library. @MarvJohnson Have I missed something here?
> 
> EDIT: if we are no longer creating the Divination Spell in obj 4, we might say something about how fetch is part of the process of how spells are created (in general)?

We're still on track to doing this! I hadn't tied the granting of items to any objective because it would likely cause regular conflicts with work others were doing in the `event.js` files. As long as we have the narrative there, the item gets granted, and the respective door(s) are only accessible with it, we should be good to go!

nmadadair commented:
> 

So is the idea in Obj 4 that the player is also "building" the Divination Spell or are the just unlocking it by completing the objective? I think either way is fine at this point, but just let me know so I can think about the narrative context a bit more.

Commit:
Updated Objective

-Updated description such that it ties in more of the story
-Updated example by replacing the immediately invoked test function with a separate definition and call

MarvelProgramming commented:
> So is the idea in Obj 4 that the player is also "building" the Divination Spell or are the just unlocking it by completing the objective? I think either way is fine at this point, but just let me know so I can think about the narrative context a bit more.

I feel like that works -- having them both build and unlocking it in Obj 4!
